src/snappy_device_agents/devices/maas2/maas_storage.py

```diff
@@ -0,0 +1,560 @@
+                and block_dev["id"] not in self.block_ids.values()
+                and partition_size <= int(block_dev["size"])
+            ):
+                return block_dev["id"]
+
+        raise MaasStorageError(
+            "No suitable block-device found for partition "
+            f"{partition_id} with size {partition_size} bytes"
+        )
+
+    def parse_block_devices(self):
+        """Find appropriate node block-device for use in layout."""
+        for partition_id, partition_size in self.partition_sizes.items():
+            self._logger_debug(f"Comparing size: Partition: {partition_size}")
+            block_device_id = self._select_block_dev(
+                partition_id, partition_size
+            )
+
+            # map partition id to block device id
+            self.block_ids[partition_id] = block_device_id
+
+    def map_block_ids(self):
+        """Map parent disks to actual node block-devices.
+
+        Updates self.device_list with "parent_disk_blkid".
+        """
+        for dev in self.device_list:
+            block_id = self.block_ids.get(dev["parent_disk"])
+            if block_id is not None:
+                dev["parent_disk_blkid"] = str(block_id)
+
+    def _validate_alloc_pct_values(self):
+        """Sanity check partition allocation percentages."""
+        alloc_pct_values = collections.defaultdict(int)
+
+        for dev in self.device_list:
+            if dev["type"] == "partition":
+                # add pct together (default to 0 if unnused)
+                alloc_pct_values[dev["parent_disk"]] += dev.get("alloc_pct", 0)
+
+        for dev_id, alloc_pct in alloc_pct_values.items():
+            if alloc_pct > 100:
+                raise MaasStorageError(
+                    "The total percentage of the partitions on disk "
+                    f"'{dev_id}' exceeds 100."
+                )
+
+    def create_partition_sizes(self):
+        """Calculate actual partition size to write to disk."""
+        self._validate_alloc_pct_values()
+
+        for dev in self.device_list:
+            if dev["type"] == "partition":
+                # find corresponding block device
+                for block_dev in self.node_info:
+                    if block_dev["id"] == self.block_ids[dev["parent_disk"]]:
+                        # get the total size of the block device in bytes
+                        total_size = int(block_dev["size"])
+                        break
+
+                if "alloc_pct" in dev:
+                    # avoid under-allocating space
+                    dev["size"] = str(
+                        math.ceil((total_size * dev.get("alloc_pct", 0)) / 100)
+                    )
+                else:
+                    if "size" not in dev:
+                        raise ValueError(
+                            f"Partition '{str(dev['id'])}' does not have an "
+                            "alloc_pct or size value."
+                        )
+                    else:
```

Comment:
This else is not needed. If you got here, then it didn't raise the exception.

Reply:
Of course. I believe I was not raising an exception here prior and then I neglected to remove the else after deciding to raise one. (or something)